### PR TITLE
Add theme and plot thickness options

### PR DESCRIPTION
**Pull request by MRog40**
I was using OpenHardwareMonitor at night on my otherwise all black OLED system, so a quick google search pointed me towards "Hey, try Libre Hardware Monitor it is still maintained". I came across an all but abandoned PR for themes and decided to revive the code and add some other additions for customization

Themes
image

Plot Stroke Thickness
image

Auto dark theme (set by windows theme setting), 2pt Stroke Thickness
image

Black theme, 2pt Stroke Thickness
image

Light theme, 4pt Stroke Thickness
image

### CODE REVIEW

**Comment by PhyxionNL** [comment]
Maybe some of the colors of the graph can be tweaked a bit? The dark blue is very hard to see on dark theme (graph is somewhat OK, but the labels are almost unreadable).

**Comment by MRog40** [comment]
These are user selected colors, I chose to have #FF0000 and #00FF00. I'm assuming if the user is editing the pen colors, we'd want to leave it even when theme is changed. The alternative would be to evaluate the brightness and invert it, maintaining hue. Thoughts on what the desired behavior would be?

Here is the auto generated colors in light and dark mode:
image

I did recognize one issue when testing this that I'm going to fix - the colors aren't updated when the theme is

**Comment by MRog40** [comment]
Updated themes:
image

And repaired the colors not being updated at the point the theme is applied

**Comment by PhyxionNL** [comment]
Thanks! One small thing, if you switch theme, the graph titlebar doesn't get updated to new color until you focus it. But I think the title bar needs to keep OS color, this is how other apps do it too.

**Comment by MRog40** [comment]
I made a video of the behavior on my end: https://imgur.com/R8r6ycM

> If you switch theme, the graph titlebar doesn't get updated to new color until you focus it.
I wasn't sure whether you meant switch OS theme, or switch LHM theme. In the video, I showed that changing both immediately updates the LHM main window and plot window themes without any additional interaction needed.

> But I think the title bar needs to keep OS color, this is how other apps do it too.
I don't think there is a proper or improper way to do this, in video I show Microsoft Edge top bar changing color with theme. Plenty of apps use their own color for the title bar, and I think this is the preferred behavior here.

**Comment by PhyxionNL** [comment]
If you switch OS theme (app preferences), the titlebar color of the plot window isn't updated until you refocus it. I tried it on two systems and both have the same problem 🤔

> I don't think there is a proper or improper way to do this, in video I show Microsoft Edge top bar changing color with theme. Plenty of apps use their own color for the title bar, and I think this is the preferred behavior here.
Edge doesn't change here either, it stays on Window' color.

**Comment by kamkie** [comment]
edge on windows 11 is system color on default theme, and user choosen otherwise (Ice mint etc)
image

**Comment by MRog40** [comment]
@PhyxionNL I unfortunately don't have a Windows 10 system to test this behavior. All I do is look in the registry for `HKEY_CURRENT_USER\Software\Microsoft\Windows\CurrentVersion\Themes\Personalize\AppsUseLightTheme` and set the theme based on that. Outside of registry behavior, I can't imagine W10 would have different UI handling for out of focus apps than W11. In terms of fixing this issue I have no ideas to try, and I don't think it should block this feature.

**Comment by pomianowski** [comment]
> @PhyxionNL I unfortunately don't have a Windows 10 system to test this behavior. All I do is look in the registry for `HKEY_CURRENT_USER\Software\Microsoft\Windows\CurrentVersion\Themes\Personalize\AppsUseLightTheme` and set the theme based on that. Outside of registry behavior, I can't imagine W10 would have different UI handling for out of focus apps than W11. In terms of fixing this issue I have no ideas to try, and I don't think it should block this feature.

You can detect the WININICHANGE (0x001A) message and then retrieve registry theme information.
https://github.com/lepoco/wpfui/blob/f2e2d136ad4aef88daa67ca17856a2fbe1abd4eb/src/Wpf.Ui/Appearance/Watcher.cs#L112
https://github.com/lepoco/wpfui/blob/f2e2d136ad4aef88daa67ca17856a2fbe1abd4eb/src/Wpf.Ui/Appearance/SystemTheme.cs#L30

Animation

**Comment by PhyxionNL** [comment]
> @PhyxionNL I unfortunately don't have a Windows 10 system to test this behavior. All I do is look in the registry for `HKEY_CURRENT_USER\Software\Microsoft\Windows\CurrentVersion\Themes\Personalize\AppsUseLightTheme` and set the theme based on that. Outside of registry behavior, I can't imagine W10 would have different UI handling for out of focus apps than W11. In terms of fixing this issue I have no ideas to try, and I don't think it should block this feature.

It's not that the detection to change theme is broken, it does that correctly. But it doesn't update the titlebar color of dialogs unless you focus them.

**Comment by JFMugen** [comment]
Awesome, thank you!

**Comment by ntoff** [comment]
If I close LHWM and reopen it, the plot window (separate window) is no longer in dark mode.

Windows 10

image

**Comment by yurisuika** [comment]
Thanks for this!

**Comment by PhyxionNL** [comment]
@MRog40 Can you address the remaining issues mentioned here so it can be merged? Thanks!

**Comment by DerekTurtleRoe** [comment]
I've been using this PR version for a long time now because I can't live without dark mode 😄 

Much appreciated @MRog40

**Comment by MRog40** [comment]
I've fixed a couple of the bugs with opening/closing and losing theme settings, still working out the title bar color

**Comment by Tamarindo94** [comment]
How can this be enabled? Theming options don't appear in my menu...

**Comment by xcasxcursex** [comment]
> How can this be enabled? Theming options don't appear in my menu...

Give the issue a thumbs up then wait and let developers develop it until one day in the releases section you see a new version and "Dark mode added" is one of the items in the change log, then shower them in praise and donations. Actually you could start praise and donations now, I guess, nobody would mind ;)

**Comment by jojo2massol** [comment]
Little issue with the mouseover : at program startup, the color of the text and the box (see image below) is a bit weird. This color is the same whatever the theme is. When the theme is changed, the issue disappears.
image

**Comment by jojo2massol** [comment]
> How can this be enabled? Theming options don't appear in my menu...

You have to compile it yourself from [MRog40:darkmode](https://github.com/MRog40/LibreHardwareMonitorDark/tree/darkmode) with Visual Studio 😉

**Comment by MRog40** [comment]
> How can this be enabled? Theming options don't appear in my menu...

Sorry for the delay everyone - day job taking priority for the moment.

I actually was previously requested to compile a release with dark mode, you can find that release here: https://github.com/LibreHardwareMonitor/LibreHardwareMonitor/releases/tag/v0.9.1

I'm going to work on getting this PR in hopefully next sprint

**Comment by DerekTurtleRoe** [comment]
@MRog40 It's fine, I am excited for it to be merged! Hopefully your job isn't too stressful or anything. Thanks for your work so far! 😄

**Comment by MRog40** [comment]
Status update:
image
It's fully working after interaction, just figuring out why it isn't working right on launch. At initial startup the plot top bar is white (on detached plot window only) and the plot details on mouse click are still blue with white text. After any interaction like clicking a checkbox on and off, it all works properly

**Comment by jojo2massol** [comment]
I hope you will be able to find out !
If you're unable to do so, maybe you can force changing theme, like with the checkbox, at startup ? Not very clean, but if it works...

**Comment by AntonioRedondo** [comment]
Thank you @MRog40 and rest of collaborators for this pull request and project. 💝 It will prevent burning my eyesight every time I check the temperature of my computer. 🫣😉

File: LibreHardwareMonitor/UI/MainForm.cs [hunk 1]
```diff
@@ -525,6 +589,29 @@ private void InitializePlotForm()
             treeView.Invalidate();
         };
 
+        _strokeThickness = new UserRadioGroup("plotStroke", 1, new[] { strokeThickness1ptMenuItem, strokeThickness2ptMenuItem, strokeThickness3ptMenuItem, strokeThickness4ptMenuItem }, _settings);
+
+        _strokeThickness.Changed += (sender, e) =>
+        {
+            switch (_strokeThickness.Value)
```
Comment (JohnMarzulli) [inline LibreHardwareMonitor/UI/MainForm.cs:596]: What about a default safe result if the value falls outside of the switch bounds?

You could also simplify this by a simple

```csharp
_plotStrokeThickness = (_strokeThickness.Value >= 0 && _strokeThickness.Value <=3)
                                                   ? _strokeThickness.Value + 1
                                                   : 4; // Or other safe value
```

**Comment by kamkie** [comment]
> Status update: image It's fully working after interaction, just figuring out why it isn't working right on launch. At initial startup the plot top bar is white (on detached plot window only) and the plot details on mouse click are still blue with white text. After any interaction like clicking a checkbox on and off, it all works properly

It also do not work when changing them is you did not interact with label befor changing theme
like plot label do not exists to set theme on it when theme is applied

**Comment by kamkie** [comment]
image
looks like OxyPlot.WindowsForms.PlotView#ShowTracker is creating it using hardcoded colors

**Comment by kamkie** [comment]
adding something like 
```c#
        ApplyTheme();
        _plot.ShowTracker(new TrackerHitResult());
        _plot.HideTracker();
        foreach (Control plotControl in _plot.Controls)
        {
            plotControl.BackColor = Theme.Current.PlotBackgroundColor;
        }
```
to PlotPanel constructor makes it working

**Comment by MRog40** [comment]
Thank you @kamkie for the fix. Two remaining issues I'll try and wrap up:
1. At initial startup the plot top bar is white (on detached plot window only).
2. The plot tracker text color is incorrect.
Interacting with the plot doesn't fix either of these, I have to interact with a sensor for it to update.

**Comment by MRog40** [comment]
I believe all the issues are now fixed. Latest was uploaded to https://github.com/MRog40/LibreHardwareMonitorDark/releases to test

File: LibreHardwareMonitor/UI/MainForm.cs (hunk 1 above)
Comment (MRog40) [reply LibreHardwareMonitor/UI/MainForm.cs]: Implemented as suggested

**Comment by MRog40** [comment]
> @MRog40 Can you address the remaining issues mentioned here so it can be merged? Thanks!

@PhyxionNL I believe all the issues have been addressed

**Comment by PhyxionNL** [comment]
> > @MRog40 Can you address the remaining issues mentioned here so it can be merged? Thanks!
> 
> @PhyxionNL I believe all the issues have been addressed

Just tried it out, but plot window is still white for me (manually set theme to Dark).

image

Also Parameters dialog is white.

**Comment by PhyxionNL** [comment]
Try this artifact: https://github.com/LibreHardwareMonitor/LibreHardwareMonitor/suites/18575006450/artifacts/1080435156
It's not working correctly for me. 
FWIW, in Personalization > Colors in Windows: I have Windows mode on Dark and app mode on Light.

**Comment by Tamarindo94** [comment]
> > How can this be enabled? Theming options don't appear in my menu...
> 
> Sorry for the delay everyone - day job taking priority for the moment.
> 
> I actually was previously requested to compile a release with dark mode, you can find that release here: https://github.com/MRog40/LibreHardwareMonitorDark/releases
> 
> I'm going to work on getting this PR in hopefully next sprint

This is great, thanks a lot!!

**Comment by MRog40** [comment]
> Try this artifact: https://github.com/LibreHardwareMonitor/LibreHardwareMonitor/suites/18575006450/artifacts/1080435156 It's not working correctly for me. FWIW, in Personalization > Colors in Windows: I have Windows mode on Dark and app mode on Light.

I've tested the provided artifact and still the theme always works for me, I've tried all combinations of separate plot window, integrated plot, and all theme options including auto and it always works. Having different Windows / App themes doesn't matter, as I only look at the "AppsUseLightTheme" registry key and never the windows settings.

**Comment by MRog40** [comment]
Here is a little video demo using your artifact showing it only responds to App theme changes, and it responds even faster than windows explorer does:

https://github.com/LibreHardwareMonitor/LibreHardwareMonitor/assets/47461109/1b63e7c3-5c7c-4e06-93d5-3da4263abfd9

**Comment by MRog40** [comment]
Merged master and uploaded another updated release here for those that are using this version: https://github.com/MRog40/LibreHardwareMonitorDark/releases

**Comment by shrddr** [comment]
Nice job [MRog40](https://github.com/MRog40)

> Merged master and uploaded another updated release here for those that are using this version: https://github.com/MRog40/LibreHardwareMonitorDark/releases

Maybe I missing something but 0.0.5.zip from here does not have the dark mode (no Themes menu either). I downloaded the darkmode branch instead and compiled, all good

**Comment by camoody1** [comment]
Can we please get this update added to a new release for all the "normies" who don't have Visual Basic installed to do that compiling or whatever other steps are needed? 😁 We just want to be able to download the main repository and get to work without having to use our brains too much. 😂

JayzTwoCents talked about your software today on his video, so I suspect you'll have a lot of traffic come through your page in the next couple of days.
https://youtu.be/sX2ThUWNXio?si=inoVJDY0Pz0ipIMh

**Comment by progenitor-amborella** [comment]
Loving the not just dark mode, but other themes from MRog40. It appears that there is an old duplicate of this at #116, which leads me to what I want to mention.

As the above comment points out, J2¢ has uploaded a video on this topic today, and I suspect many people unfamiliar with git/GitHub will be stopping by (welcome, everyone!) I would expect possible duplicate postings on the issues he brought up in his video, dark mode being one of them.

To those new, remember to be kind to the maintainers!

P.S. @camoody1 could you edit out the tracking section of your YouTube link? Everything from "?" onwards can always be deleted. :)

**Comment by MRog40** [comment]
Merged latest master into another release:
https://github.com/MRog40/LibreHardwareMonitorDark/releases/tag/0.0.6

**Comment by KyodaiKen** [comment]
> Merged latest master into another release: https://github.com/MRog40/LibreHardwareMonitorDark/releases/tag/0.0.6

Good job man! :)

**Comment by PhyxionNL** [comment]
Let's run the PR again to see if it's working correctly now.

**Comment by PhyxionNL** [comment]
Still not working correctly @MRog40 .
To reproduce:
- Set LHM to light and make sure "Show plot" is disabled.
- Close and open LHM again.
- Now first change theme to Dark and then enable "Show plot". The plot window is now white.

The rest seems to work OK, if this is fixed I can merge the PR.

**Comment by kamkie** [comment]
Maybe this little corner case can be left as a bug for a later fix?

**Comment by PhyxionNL** [comment]
> Maybe this little corner case can be left as a bug for a later fix?

Let's fix it right properly.

**Comment by xmikedanielsx** [comment]
> Still not working correctly @MRog40 . To reproduce:
> 
> * Set LHM to light and make sure "Show plot" is disabled.
> * Close and open LHM again.
> * Now first change theme to Dark and then enable "Show plot". The plot window is now white.
> 
> The rest seems to work OK, if this is fixed I can merge the PR.

Can confirm. Just tested this and can reproduce.

**Comment by Tellinq** [comment]
> Still not working correctly @MRog40 . To reproduce:
> 
> * Set LHM to light and make sure "Show plot" is disabled.
> * Close and open LHM again.
> * Now first change theme to Dark and then enable "Show plot". The plot window is now white.
> 
> The rest seems to work OK, if this is fixed I can merge the PR.

Can also verify that I can reproduce via these steps
image

image



> @MRog40 @PhyxionNL To fix this issue. 1 Line needs to be added to MainForm.cs
> 
> ```cs
> Theme.Current.Apply(_plotForm);
> ```
> 
> image

Can also verify that if I add this line I'm no longer able to replicate this issue anymore

**Comment by xmikedanielsx** [comment]
It adds extra overhead to the app when opening the plot (negligible but overhead none the less) but ensures that it is always grabbing current theme.

**Comment by MRog40** [comment]
@PhyxionNL Thank you for repro steps! Until now I could not figure out what the issue was. Fixed.

**Comment by xmikedanielsx** [comment]
@MRog40  Thanks Mate
@PhyxionNL  Let's get this merged! 
Thanks to everyone for testing and moving things forward!

**Pull request merged**